File: phonetics/embed_vowel_test.c
/* Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "phonetics/embed_vowel.h"

#include <math.h>
#include <stdlib.h>
#include <string.h>

#include "dsp/logging.h"
#include "dsp/read_wav_file.h"
#include "frontend/carl_frontend.h"

#define kBlockSize 64

/* Tests that target lookup functions work. */
void TestTargetLookup() {
  puts("TestTargetLookup");
  int i;
  for (i = 0; i < kEmbedVowelNumTargets; ++i) {
    /* Target should be closest to itself. */
    CHECK(EmbedVowelClosestTarget(kEmbedVowelTargets[i].coord) == i);
    /* Lookup by target name. */
    CHECK(EmbedVowelTargetByName(kEmbedVowelTargets[i].name) == i);
  }

  /* Searching for a nonexistent target name. */
  CHECK(EmbedVowelTargetByName("nonsense") == -1);
}

/* Runs CarlFrontend + EmbedVowel on a short WAV recording of a pure phone, and
 * checks that the output coordinate is usually near the correct target.
 */
void TestPhone(const char* phone) {
  printf("TestPhone(\"%s\")\n", phone);
  char wav_file[1024];
  sprintf(wav_file, "testdata/phone_%s.wav", phone);

  size_t num_samples;
  int num_channels;
  int sample_rate_hz;
  int16_t* input_int16 = (int16_t*)CHECK_NOTNULL(Read16BitWavFile(
      wav_file, &num_samples, &num_channels, &sample_rate_hz));
  float* frame = (float*)CHECK_NOTNULL(
      malloc(sizeof(float) * kEmbedVowelNumChannels));
  CHECK(num_channels == 1);
  CHECK(sample_rate_hz == 16000);

  CarlFrontendParams frontend_params = kCarlFrontendDefaultParams;
  frontend_params.block_size = kBlockSize;
  CarlFrontend* frontend = CHECK_NOTNULL(CarlFrontendMake(&frontend_params));
  CHECK(CarlFrontendNumChannels(frontend) == kEmbedVowelNumChannels);

  const int intended_target = EmbedVowelTargetByName(phone);
  int count_correct = 0;
  int count_total = 0;

  int start;
  int i;
  for (start = 0; start + kBlockSize < num_samples; start += kBlockSize) {
    float input_float[kBlockSize];
    for (i = 0; i < kBlockSize; ++i) {
      input_float[i] = input_int16[start + i] / 32768.0f;
    }

    CarlFrontendProcessSamples(frontend, input_float, frame);
    float coord[2];
    EmbedVowel(frame, coord);

    count_correct += (EmbedVowelClosestTarget(coord) == intended_target);
    ++count_total;
  }

  CarlFrontendFree(frontend);
  free(frame);
  free(input_int16);

  const float accuracy = (float)count_correct / count_total;
  CHECK(accuracy >= 0.7f);
}

int main(int argc, char** argv) {
  TestTargetLookup();
  TestPhone("aa");
  TestPhone("uw");
  TestPhone("ih");
  TestPhone("iy");
  TestPhone("eh");
  TestPhone("ae");
  TestPhone("uh");
  TestPhone("er");

  puts("PASS");
  return EXIT_SUCCESS;
}

File: extras/python/phonetics/embed_vowel_test.py
# Copyright 2019 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may not
# use this file except in compliance with the License. You may obtain a copy of
# the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations under
# the License.

"""Tests for vowel embedding Python bindings."""

import unittest
import numpy as np

from extras.python import dsp
from extras.python import frontend
from extras.python.phonetics import embed_vowel


class EmbedVowelTest(unittest.TestCase):

  def _run_phone(self, phone):
    self.assertEqual(len(embed_vowel.TARGET_NAMES), embed_vowel.NUM_TARGETS)
    self.assertEqual(len(embed_vowel.TARGET_COORDS), embed_vowel.NUM_TARGETS)
    phone_index = embed_vowel.TARGET_NAMES.index(phone)

    wav_file = (
        f'extras/test/testdata/phone_{phone}.wav')
    samples, _ = dsp.read_wav_file(wav_file, dtype=np.float32)
    samples = samples.mean(axis=1)

    carl = frontend.CarlFrontend()
    self.assertEqual(carl.num_channels, embed_vowel.NUM_CHANNELS)
    samples = samples[:len(samples) - len(samples) % carl.block_size]
    frames = carl.process_samples(samples)
    coords = embed_vowel.embed_vowel(frames)

    distance_from_targets = np.linalg.norm(
        coords[:, np.newaxis, :]
        - embed_vowel.TARGET_COORDS[np.newaxis, :, :],
        axis=-1)

    # Compare `coords` with embed_vowel_scores.
    scores = embed_vowel.embed_vowel_scores(frames)
    self.assertEqual(scores.shape, (frames.shape[0], embed_vowel.NUM_TARGETS))
    np.testing.assert_array_equal(scores.argmax(axis=1),
                                  distance_from_targets.argmin(axis=1))
    np.testing.assert_allclose(
        scores, np.exp(-4.0 * distance_from_targets), atol=1e-6)

    # Compute L2 distance from intended_target.
    distance_from_intended = distance_from_targets[:, phone_index]
    # Mean distance from the intended target is not too large.
    self.assertLessEqual(distance_from_intended.mean(), 0.35)

    other = [j for j in range(embed_vowel.NUM_TARGETS) if j != phone_index]
    min_distance_from_other = distance_from_targets[:, other].min(axis=1)
    accuracy = float(np.count_nonzero(
        distance_from_intended < min_distance_from_other)) / len(frames)
    # Coordinate is closest to the indended target >= 70% of the time.
    self.assertGreaterEqual(accuracy, 0.7)

  def test_phone_aa(self):
    self._run_phone('aa')

  def test_phone_uw(self):
    self._run_phone('uw')

  def test_phone_er(self):
    self._run_phone('er')

  def test_phone_ih(self):
    self._run_phone('ih')

  def test_phone_iy(self):
    self._run_phone('iy')

  def test_phone_eh(self):
    self._run_phone('eh')

  def test_phone_ae(self):
    self._run_phone('ae')

  def test_phone_uh(self):
    self._run_phone('uh')


if __name__ == '__main__':
  unittest.main()
